### DISCUSSION TITLE
Is there function in AiQueue that mimics AI travel behaviour?

### DISCUSSION THREAD

**Discussion by PawelPerek** [Try/OpenGothic#505] [Q&A] [answered]
Hi there, 
I see that `aigoto` from Marvin mode is using `AiQueue::aiGoToPoint` which is basically only traversing through waypoint graph (which is good because that how it works in original Gothic and Gothic II). 
However I was wondering if there is a function that does the same thing as what travelling companion does, i.e. traveling through graph and fighting when it detects enemy mob in some radius (and perhaps loots like Alligator Jack in Jarkendar, but It's probably just hacky animation)? I know it's obviously possible because the game does this but not sure if it's implemented inside game scripts or engine itself.

P.S. The code of the engine is joy to look though, kudos @Try for keeping the codebase in such high-quality state. I've seen commercial apps code in much worse shape than this one.

**Comment by Try** [comment, marked as answer]
Hi, @PawelPerek !

> if it's implemented inside game scripts or engine itself.

Yes, go to point and do activities while on that is quite high-level behavior, controlled by scripts. I can give a high level overview of what building block are in place:

1. Npc has a "state" state is triplet of script functions, for example `ZS_Stand_ArmsCrossed` + `ZS_Stand_ArmsCrossed_loop` + `ZS_Stand_ArmsCrossed_end`. 
2. Npc do have a time-table. Every entry in time-table is state+waypoint. for example `ZS_Stand` + `FP_GATE_GUARD_01`.
Time-table may have only single entry, when npc need to stay in one place.
3. Npc has perceptions system to detect player/enemies/doors.

All this has to work together to achieve npc travel. Example, when Xardas goes to help with ritual, script does:
1. `Npc_ExchangeRoutine(self,"RitualInnosEyeRepair");` <-- assign new time-table
```c++
func void Rtn_RitualInnosEyeRepair_100()
{
	TA_Stand_ArmsCrossed(8,0,23,0,"NW_TROLLAREA_RITUAL_01"); // ZS_STAND at point "NW_TROLLAREA_RITUAL_01" 
	TA_Stand_ArmsCrossed(23,0,8,0,"NW_TROLLAREA_RITUAL_01");
};
``` 
2. In stand routine:
```
func void ZS_Stand_ArmsCrossed()
{
	Perception_Set_Normal();
	B_ResetAll(self);
	AI_SetWalkMode(self,NPC_WALK);
	if(Npc_GetDistToWP(self,self.wp) > TA_DIST_SELFWP_MAX)
	{
		AI_GotoWP(self,self.wp); // go-go to waypoint
	};
	self.aivar[AIV_TAPOSITION] = NOTINPOS;
};
```
3. While he on a go, if monster appears he with attack and change state to `ZS_ATTACK`
4. When fight is over - return to state defined in time-table and continue walk to "NW_TROLLAREA_RITUAL_01"

**Reply by PawelPerek** [reply]
Thanks, that sounds like someting not terribly hard to reimplement. I am not too familiar with Gothic architecture, are those daedalus scripts? Are there some reverse-engineered version of game scripts on github (just like you send)?

**Comment by Try** [comment]
>that sounds like someting not terribly hard to reimplement.

Should "just work" already, isn't it?

> Are there some reverse-engineered version of game scripts on github (just like you send)?

Don't remember where exactly I find it, but generally there are plenty: https://github.com/IDizor/GothicScripts and easy to decompile.

**Reply by PawelPerek** [reply]
I think so, but there is an overwhelming amount of tools and information about gothic modding (how to compile .d scripts, what is daedalus/ikarus/ninja/union/probably other stuff I never heard about) that I don't have cognitive load to deal with. At the end of the day I'm just doing it to cleanup Khorinis from mobs at the end of chapter 1 because it is the most tedious part of my annual G1/G2 playthough, so I think baking perception and ZS_Attack system for pc_hero directly into the engine will be the fastest way to do that (however insane that sounds 😝). Anyways, thanks for help!
